File: Android/shiba/src/main/java/moe/tlaster/shiba/parser/ShibaParserWrapper.kt
package moe.tlaster.shiba.parser

import moe.tlaster.shiba.type.*
import org.w3c.dom.Attr
import org.w3c.dom.Element
import org.w3c.dom.Node
import org.xml.sax.InputSource
import java.io.StringReader
import javax.xml.parsers.*

private val Node.name
    get() = this.localName
private val Node.namespaceName
    get() = this.namespaceURI

private fun Node.hasChildElement(): Boolean {
    if (!this.hasChildNodes()) {
        return false
    }
    for (index in 0 until this.childNodes.length) {
        val item = this.childNodes.item(index)
        if (item is Element) {
            return true
        }
    }
    return false
}

private fun Node.childElements(): List<Element> {
    if (!this.hasChildElement()) {
        return listOf()
    }
    return (0 until childNodes.length).map { index ->
        childNodes.item(index)
                ?.takeIf {
                    it is Element
                }?.let {
                    it as Element
                }
    }.filter {
        it != null
    }.map {
        it!!
    }
}

interface IShibaVisitor {
    val type: Class<*>
    val returnType: Class<*>
    fun visit(tree: Any): Any
}

private val visitors = ArrayList<IShibaVisitor>().apply {
    add(ViewVisitor())
    add(ElementPropertyVisitor())
    add(PropertyVisitor())
    add(ShibaTokenVisitor())
    add(ShibaMapVisitor())
    add(FunctionVisitor())
    add(ShibaExtensionVisitor())
}

private inline fun <reified T> Any.visit(): T? {
    return visitors.find { it.type == this.javaClass && it.returnType == T::class.java }?.visit(this) as T?
}

private abstract class AbsVisitor<T : Any, K : Any> : IShibaVisitor {
    override fun visit(tree: Any): Any {
        return parse(tree as T)
    }

    abstract fun parse(tree: T): K
}

private final class ViewVisitor(override val type: Class<*> = Element::class.java, override val returnType: Class<*> = View::class.java) : AbsVisitor<Element, View>() {
    override fun parse(tree: Element): View {
        val viewName = (tree as Node).visit<ShibaToken>()
                ?: throw IllegalArgumentException("view name can not be null at $tree")
        val view = View(viewName)
        for (index in 0 until tree.attributes.length) {
            tree.attributes.item(index)?.let {
                it as Attr
            }?.let {
                it.visit<Property>()
            }?.let {
                view.properties += it
            }
        }
        if (!tree.hasChildElement()) {
            // TODO: default property
        } else {
            tree.childElements().forEach { element ->
                if (element.localName.contains('.')
                        && element.localName.split('.').firstOrNull() == viewName.value
                        && element.namespaceName == viewName.prefix) {
                    element.visit<Property>()?.let {
                        view.properties += it
                    }
                } else {
                    element.visit<View>()?.let {
                        view.children += it
                    }
                }
            }
        }

        return view
    }
}

private fun parsePropertyValue(propertyValue: String): Any {
    val OpenCurly = '{'
    val CloseCurly = '}'
    val OpenBracket = '['
    val CloseBracket = ']'
    val ExtensionStart = '$'
    val value = propertyValue.trim()
    when {
        value.startsWith(OpenCurly) && value.endsWith(CloseCurly) -> {
            val subValue = value.substring(1, value.length - 2)
            val function = subValue.visit<ShibaFunction>()
            if (function != null) {
                return function
            }
        }
        value.startsWith(OpenBracket) && value.endsWith(CloseBracket) -> {
            val subValue = value.substring(1, value.length - 2)
            val map = subValue.visit<ShibaMap>()
            if (map != null) {
                return map
            }
        }
        value.startsWith(ExtensionStart) -> {
            val extension = value.visit<ShibaExtension>()
            if (extension != null) {
                return extension
            }
        }
    }

    if (value == "true" || value == "false") {
        return value.toBoolean()
    }


    val numberValue = value.toBigDecimalOrNull()
    if (numberValue != null) {
        return numberValue
    }
    return value
}

private final class ElementPropertyVisitor(override val type: Class<*> = Element::class.java, override val returnType: Class<*> = Property::class.java) : AbsVisitor<Element, Property>() {
    override fun parse(tree: Element): Property {
        val name = (tree as Node).visit<ShibaToken>()
                ?: throw IllegalArgumentException("view name can not be null at $tree")
        if (tree.hasChildElement()) {
            tree.childElements().firstOrNull()?.visit<View>()?.let {
                return Property(name, it)
            }
        }
        return Property(name, parsePropertyValue(tree.textContent))
    }
}

private final class PropertyVisitor(override val type: Class<*> = Attr::class.java, override val returnType: Class<*> = Property::class.java) : AbsVisitor<Attr, Property>() {
    override fun parse(tree: Attr): Property {
        val tokenName = (tree as Node).visit<ShibaToken>()
                ?: throw IllegalArgumentException("view name can not be null at $tree")
        val value = parsePropertyValue(tree.value)
        return Property(tokenName, value)
    }
}


private final class ShibaExtensionVisitor(override val type: Class<*> = String::class.java, override val returnType: Class<*> = ShibaExtension::class.java) : AbsVisitor<String, ShibaExtension>() {
    private val ScriptStart = '{'
    private val ExtensionStart = '$'
    override fun parse(tree: String): ShibaExtension {
        if (!tree.startsWith(ExtensionStart)) {
            throw IllegalArgumentException("wrong shiba extension at $tree")
        }
        var value = tree.trimStart(ExtensionStart)
        var index = tree.indexOf(' ')
        val name = value.substring(0, index)
        value = value.substring(index + 1, value.length - index - 1)
        index = value.indexOf(ScriptStart)
        if (index == -1) {
            return ShibaExtension(name.trim(), value.trim(), null)
        }
        val bindingValue = value.substring(0, index).trim()
        value = value.substring(index + 1, value.length - index - 1).trim()
        val script = value.substring(1, value.length - 2)
        return ShibaExtension(name.trim(), bindingValue, script)
    }

}

private final class ShibaTokenVisitor(override val type: Class<*> = Node::class.java, override val returnType: Class<*> = ShibaToken::class.java) : AbsVisitor<Node, ShibaToken>() {
    override fun parse(tree: Node): ShibaToken {
        return if (tree.prefix.isNullOrEmpty()) {
            ShibaToken("", tree.name)
        } else {
            ShibaToken(tree.namespaceName, tree.name)
        }
    }
}

private final class ShibaMapVisitor(override val type: Class<*> = String::class.java, override val returnType: Class<*> = ShibaMap::class.java) : AbsVisitor<String, ShibaMap>() {
    private val EqualSign = '='
    private val Comma = ','
    override fun parse(tree: String): ShibaMap {
        return ShibaMap(tree
                .split(Comma)
                .map { it.trim() }
                .map { it.split(EqualSign) }
                .map { it.first() to it.last() }
                .toMap())
    }
}

private final class FunctionVisitor(override val type: Class<*> = String::class.java, override val returnType: Class<*> = ShibaFunction::class.java) : AbsVisitor<String, ShibaFunction>() {
    private val Comma = ','
    override fun parse(tree: String): ShibaFunction {
        val index = tree.indexOf('(')
        val name = tree.substring(0, index)
        val function = ShibaFunction(name.trim())
        val value = tree.substring(index + 1, tree.length - index - 2)
        val param = value.split(Comma).map {
            parsePropertyValue(it)
        }
        function.parameter += param
        return function
    }
}


class ShibaParserWrapper {
    private fun parseGrammarTree(input: String): Element? {
        val dom = DocumentBuilderFactory.newInstance().apply {
            isNamespaceAware = true
        }.newDocumentBuilder().parse(InputSource(StringReader(input)))
        return dom.documentElement
    }

    fun parse(input: String): View? {
        val tree = parseGrammarTree(input)
        return tree?.visit<View>()
    }
}

File: UWP/Shiba/Parser/ShibaParserWrapper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Xml.Linq;
using Shiba.Controls;
using Shiba.Internal;

namespace Shiba.Parser
{
    internal interface IShibaVisitor
    {
        Type HandleType { get; }
        Type ReturnType { get; }
        object Visit(object tree);
    }

    internal abstract class ShibaVisitor : IShibaVisitor
    {
        private static readonly List<IShibaVisitor> Visitors;

        static ShibaVisitor()
        {
            Visitors = typeof(IShibaVisitor)
                .Assembly
                .DefinedTypes
                .Where(it =>
                    it.IsClass && !it.IsAbstract &&
                    typeof(IShibaVisitor).GetTypeInfo().IsAssignableFrom(it.GetTypeInfo()))
                .Select(it => Activator.CreateInstance(it) as IShibaVisitor)
                .ToList();
        }

        public abstract Type HandleType { get; }
        public abstract Type ReturnType { get; }
        public abstract object Visit(object tree);

        //protected internal static object GetValue(object tree)
        //{
        //    if (tree == null) return null;
        //    var visitor = Visitors.FirstOrDefault(it => it.HandleType == tree.GetType());
            
        //    return visitor.Visit(tree);
        //}

        protected internal static T GetValue<T>(object tree)
        {
            if (tree == null) return default;
            var visitor = Visitors.FirstOrDefault(it => it.HandleType == tree.GetType() && it.ReturnType == typeof(T));
            if (visitor == null) return default;
            var value = visitor.Visit(tree);

            if (value is T result) return result;

            return default;
        }
    }

    internal abstract class GenericVisitor<T, K> : ShibaVisitor
    {
        public override Type HandleType { get; } = typeof(T);
        public override Type ReturnType { get; } = typeof(K);

        public override object Visit(object tree)
        {
            if (tree is T t)
            {
                return Parse(t);
            }

            return null;
        }

        protected abstract K Parse(T tree);
    }

    internal sealed class ViewVisitor : GenericVisitor<XElement, View>
    {
        protected override View Parse(XElement tree)
        {
            var viewName = GetValue<ShibaToken>(tree.Name);
            var view = new View(viewName, tree.ToString());
            
            view.Properties.AddRange(tree.Attributes().Select(GetValue<Property>));

            if (!tree.HasElements && !string.IsNullOrEmpty(tree.Value))
            {
                view.DefaultValue = Singleton<ValueParser>.Instance.Parse(tree.Value);
            }

            var propertyElement = tree.Elements()
                .Where(it =>
                    it.Name.LocalName.Contains(".") &&
                    it.Name.LocalName.Split('.').FirstOrDefault() == viewName.Value &&
                    it.Name.NamespaceName == viewName.Prefix).ToList();
            view.Properties.AddRange(propertyElement
                .Select(GetValue<Property>));

            view.Children.AddRange(tree.Elements().Except(propertyElement).Select(it =>
            {
                var child = GetValue<View>(it);
                child.Parent = view;
                return child;
            }));

            return view;
        }
    }

    internal sealed class PropertyElementVisitor : PropertyBaseVisitor<XElement>
    {
        protected override Property Parse(XElement tree)
        {
            var name = GetValue<ShibaToken>(tree.Name);
            if (tree.HasElements)
            {
                return new Property(name, GetValue<View>(tree.Elements().FirstOrDefault()));
            }
            else
            {
                return new Property(name, GetPropertyValue(tree.Value));
            }
        }
    }

    internal sealed class ValueParser
    {
        private const char OpenCurly = '{';
        private const char CloseCurly = '}';
        private const char OpenBracket = '[';
        private const char CloseBracket = ']';
        private const char ExtensionStart = '$';
        public object Parse(string value)
        {
            value = value.Trim();
            if (value.StartsWith(OpenCurly) && value.EndsWith(CloseCurly))
            {
                var subValue = value.Substring(1, value.Length - 2);
                var function = ShibaVisitor.GetValue<ShibaFunction>(subValue);
                if (function != null)
                {
                    return function;
                }
            }
            else if (value.StartsWith(OpenBracket) && value.EndsWith(CloseBracket))
            {
                var subValue = value.Substring(1, value.Length - 2);
                var map = ShibaVisitor.GetValue<ShibaMap>(subValue);
                if (map != null)
                {
                    return map;
                }
            }
            else if (value.StartsWith(ExtensionStart))
            {
                var extension = ShibaVisitor.GetValue<ShibaExtension>(value);
                if (extension != null)
                {
                    return extension;
                }
            }

            if (value.TryChangeType(typeof(bool), out var boolValue))
            {
                return boolValue;
            }
            if (value.TryChangeType(typeof(decimal), out var numberValue))
            {
                return numberValue;
            }
            return value;
        }
    }

    internal abstract class PropertyBaseVisitor<T> : GenericVisitor<T, Property>
    {

        protected object GetPropertyValue(string value)
        {
            return Singleton<ValueParser>.Instance.Parse(value);
        }
    }

    internal sealed class PropertyVisitor : PropertyBaseVisitor<XAttribute>
    {
        protected override Property Parse(XAttribute tree)
        {
            var tokenName = GetValue<ShibaToken>(tree.Name);
            return new Property(tokenName, GetPropertyValue(tree.Value));
        }
    }

    internal sealed class ShibaTokenVisitor : GenericVisitor<XName, ShibaToken>
    {
        protected override ShibaToken Parse(XName tree)
        {
            if (tree.LocalName.Contains('.'))
            {
                return new ShibaToken(tree.LocalName.Split('.').Skip(1).FirstOrDefault());
            }
            if (string.IsNullOrEmpty(tree.NamespaceName))
            {
                return new ShibaToken(tree.LocalName);
            }
            return new ShibaToken(tree.NamespaceName, tree.LocalName);
        }
    }

    internal sealed class ShibaMapVisitor : GenericVisitor<string, ShibaMap>
    {
        private const char EqualSign = '=';
        private const char Comma = ',';
        protected override ShibaMap Parse(string tree)
        {
            return new ShibaMap(tree
                    .Split(Comma)
                    .Select(it => it.Trim())
                    .Select(it => it.Split(EqualSign))
                    .ToDictionary(it => it.FirstOrDefault(), it => it.Skip(1).FirstOrDefault()));
        }
    }

    internal sealed class FunctionVisitor : GenericVisitor<string, ShibaFunction>
    {
        private const char Comma = ',';
        protected override ShibaFunction Parse(string value)
        {
            var index = value.IndexOf('(');
            var name = value.Substring(0, index);
            var function = new ShibaFunction(name.Trim());
            value = value.Substring(index + 1, value.Length - index - 2);
            var param = value.Split(Comma).Select(it => Singleton<ValueParser>.Instance.Parse(it.Trim()));
            function.Parameters.AddRange(param);
            return function;
        }
    }

    internal sealed class ShibaExtensionVisitor : GenericVisitor<string, ShibaExtension>
    {
        private const char ScriptStart = '{';
        private const char ExtensionStart = '$';

        protected override ShibaExtension Parse(string tree)
        {
            if (!tree.StartsWith(ExtensionStart))
            {
                return null;
            }

            // Checking for extension name
            var value = tree.TrimStart(ExtensionStart);
            var index = value.IndexOf(' ');
            var name = value.Substring(0, index);
            // Checking for extension value
            value = value.Substring(index + 1, value.Length - index - 1);
            index = value.IndexOf(ScriptStart);
            if (index == -1)// value only contains binding like `$bind path`
            {
                return new ShibaExtension(name.Trim(), value.Trim(), string.Empty);
            }

            var bindingValue = value.Substring(0, index).Trim();
            // or value contains script like `$bind path, {return hello}`
            value = value.Substring(index + 1, value.Length - index - 1).Trim();
            var script = value.Substring(1, value.Length - 2);
            return new ShibaExtension(name.Trim(), bindingValue, script);
        }
    }

    public class ShibaParserWrapper
    {
        public View Parse(string input)
        {
            var tree = ParseGrammarTree(input);
            return ShibaVisitor.GetValue<View>(tree);
        }

        private XElement ParseGrammarTree(string input)
        {
            return XDocument.Parse(input).Root;
        }
    }
}